File: app/src/main/java/padev/badhabits/Data/CRUD/HabitDetailsCRUD.java
package padev.badhabits.Data.CRUD;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;

import padev.badhabits.Data.AbstractData;
import padev.badhabits.Data.HabitDetails;
import padev.badhabits.Data.IDataAccessObject;

public class HabitDetailsCRUD extends SQLiteOpenHelper implements IDataAccessObject, CRUDSingleton {

    private static HabitDetailsCRUD instance = null;

    private static final int DATABASE_VERSION = 1;
    private static final String DATABASE_NAME = "badHabitsManager";
    private static final String TABLE_HABITS_DETAILS = "habitsDetails";
    private static final String KEY_ID = "id";
    private static final String KEY_HABIT_ID = "habit_id";
    private static final String KEY_DOSE = "dose";
    private static final String KEY_CONCENTRATION = "concentration";
    private static final String KEY_WEIGHT = "weight";


    private HabitDetailsCRUD(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);

        onUpgrade(getWritableDatabase(), 1, 1);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        String CREATE_HABIT_DETAILS_TABLE = "CREATE TABLE " + TABLE_HABITS_DETAILS + "("
                + KEY_ID + " INTEGER PRIMARY KEY," + KEY_HABIT_ID + " INTEGER," + KEY_DOSE + " INTEGER," + KEY_CONCENTRATION +  " INTEGER," + KEY_WEIGHT +  " INTEGER" + ")";
        //db.execSQL("PRAGMA foreign_keys=ON;");
        db.execSQL(CREATE_HABIT_DETAILS_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE IF EXISTS " + TABLE_HABITS_DETAILS);

        onCreate(db);
    }

    @Override
    public void insertData(AbstractData data) {

        HabitDetails habitDetails = (HabitDetails) data;

        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put(KEY_HABIT_ID, habitDetails.getHabitId());
        values.put(KEY_DOSE, habitDetails.getDose());
        values.put(KEY_CONCENTRATION, habitDetails.getConcentration());
        values.put(KEY_WEIGHT, habitDetails.getWeight());

        habitDetails.setId(db.insert(TABLE_HABITS_DETAILS, null, values));

        db.close();
    }

    @Override
    public AbstractData selectData(int id) {
        SQLiteDatabase db = this.getReadableDatabase();

        Cursor cursor = db.query(TABLE_HABITS_DETAILS, new String[] { KEY_ID,
                        KEY_DOSE, KEY_CONCENTRATION, KEY_WEIGHT}, KEY_ID + "=?",
                new String[] { String.valueOf(id) }, null, null, null, null);

        if (cursor != null){
            cursor.moveToFirst();
        }

        assert cursor != null;

        return new HabitDetails(Integer.parseInt(cursor.getString(0)), Integer.parseInt(cursor.getString(1)), Integer.parseInt(cursor.getString(2)), Integer.parseInt(cursor.getString(3)));
    }

    public ArrayList<AbstractData> selectDataByHabitId(long habitId) {

        ArrayList<AbstractData> habitDetails = new ArrayList<>();

        SQLiteDatabase db = this.getReadableDatabase();

        Cursor cursor = db.query(TABLE_HABITS_DETAILS, null, KEY_HABIT_ID + " = ?", new String[]{String.valueOf(habitId)}, null, null,null);

        if (cursor.moveToFirst()) {
            do {
                HabitDetails habitDetail = new HabitDetails(Integer.parseInt(cursor.getString(0)), Integer.parseInt(cursor.getString(1)), Integer.parseInt(cursor.getString(2)), Integer.parseInt(cursor.getString(3)), Integer.parseInt(cursor.getString(4)));
                                                                                                                                                                                                           habitDetails.add(habitDetail);
            } while (cursor.moveToNext());
        }

        return habitDetails;
    }

    @Override
    public ArrayList<AbstractData> selectAllData() {

        // TODO if you want to
        return new ArrayList<>();
    }

    @Override
    public int updateData(AbstractData data) {
        HabitDetails habitDetails = (HabitDetails) data;

        SQLiteDatabase db = this.getWritableDatabase();

        ContentValues values = new ContentValues();
        values.put(KEY_DOSE, habitDetails.getDose());
        values.put(KEY_CONCENTRATION, habitDetails.getConcentration());
        values.put(KEY_WEIGHT, habitDetails.getWeight());

        return db.update(TABLE_HABITS_DETAILS, values, KEY_ID + " = ?",
                new String[] { String.valueOf(habitDetails.getId()) });
    }

    @Override
    public void deleteData(int id) {
        SQLiteDatabase db = this.getWritableDatabase();
        db.delete(TABLE_HABITS_DETAILS, KEY_ID + " = ?", new String[] { String.valueOf(id) });
        db.close();
    }

    @Override
    public void deleteAllData() {
        SQLiteDatabase db = this.getWritableDatabase();
        db.delete(TABLE_HABITS_DETAILS, null, null);
        db.close();
    }

    public static HabitDetailsCRUD getInstance(Context context) {
        if (instance == null) instance = new HabitDetailsCRUD(context);

        return instance;
    }
}

File: app/src/main/java/padev/badhabits/Data/CRUD/HabitDetailsCRUD.kt
package padev.badhabits.Data.CRUD;

import android.annotation.SuppressLint
import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.ArrayList;

import padev.badhabits.Data.AbstractData;
import padev.badhabits.Data.HabitDetails;
import padev.badhabits.Data.IDataAccessObject;

class HabitDetailsCRUD(context: Context): SQLiteOpenHelper(context, "padev.badhabits.db", null, 1), IDataAccessObject {

/*
    private val DATABASE_VERSION = 1
    private val DATABASE_NAME = "badHabitsManager"
*/

    private val TABLE_HABITS_DETAILS = "habitsDetails"
    private val KEY_ID = "id"
    private val KEY_HABIT_ID = "habit_id"
    private val KEY_DOSE = "dose"
    private val KEY_CONCENTRATION = "concentration"
    private val KEY_WEIGHT = "weight"


    init {
        writableDatabase.execSQL("CREATE TABLE IF NOT EXISTS $TABLE_HABITS_DETAILS($KEY_ID INTEGER PRIMARY KEY,$KEY_HABIT_ID INTEGER,$KEY_DOSE INTEGER,$KEY_CONCENTRATION INTEGER,$KEY_WEIGHT INTEGER)")
    }

    override fun onCreate(db: SQLiteDatabase) {
        //db.execSQL("PRAGMA foreign_keys=ON;");
        db.execSQL("CREATE TABLE $TABLE_HABITS_DETAILS($KEY_ID INTEGER PRIMARY KEY,$KEY_HABIT_ID INTEGER,$KEY_DOSE INTEGER,$KEY_CONCENTRATION INTEGER,$KEY_WEIGHT INTEGER)")
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        db.execSQL("DROP TABLE IF EXISTS $TABLE_HABITS_DETAILS")

        onCreate(db)
    }

    override fun insertData(data: AbstractData) {

        val habitDetails = data as HabitDetails

        val db = this.writableDatabase
        val values = ContentValues()
        values.put(KEY_HABIT_ID, habitDetails.habitId)
        values.put(KEY_DOSE, habitDetails.dose)
        values.put(KEY_CONCENTRATION, habitDetails.concentration)
        values.put(KEY_WEIGHT, habitDetails.weight)

        habitDetails.id = db.insert(TABLE_HABITS_DETAILS, null, values)

        db.close()
    }

    @SuppressLint("Recycle")
    override fun selectData(id: Int): AbstractData {
        val db = this.readableDatabase

        val cursor = db.query(TABLE_HABITS_DETAILS, arrayOf(KEY_ID,
                        KEY_DOSE, KEY_CONCENTRATION, KEY_WEIGHT), "$KEY_ID=?",
                arrayOf(id.toString()), null, null, null, null)

        cursor?.moveToFirst()

        return HabitDetails(cursor.getString(0).toLong(),
                cursor.getString(1).toInt(),
                cursor.getString(2).toInt(),
                cursor.getString(3).toInt())
    }

    fun selectDataByHabitId(habitId: Long): ArrayList<AbstractData>  {

        val habitDetails = ArrayList<AbstractData>()

        val db = this.readableDatabase

        val cursor = db.query(TABLE_HABITS_DETAILS, null, "$KEY_HABIT_ID = ?", arrayOf(habitId.toString()), null, null,null)

        if (cursor.moveToFirst()) {
            do {
                val habitDetail = HabitDetails(
                        cursor.getString(0).toLong(),
                        cursor.getString(1).toLong(),
                        cursor.getString(2).toInt(),
                        cursor.getString(3).toInt(),
                        cursor.getString(4).toInt())
                                                                                                                                                                                                           habitDetails.add(habitDetail);
            } while (cursor.moveToNext())
        }

        return habitDetails
    }

    override fun selectAllData(): ArrayList<AbstractData> {

        // TODO if you want to
        return ArrayList()
    }

    override fun updateData(data: AbstractData): Int {
        val habitDetails = data as HabitDetails

        val db = this.writableDatabase

        val values = ContentValues()
        values.put(KEY_DOSE, habitDetails.dose)
        values.put(KEY_CONCENTRATION, habitDetails.concentration)
        values.put(KEY_WEIGHT, habitDetails.weight)

        return db.update(TABLE_HABITS_DETAILS, values, "$KEY_ID = ?",
                arrayOf(habitDetails.id.toString()))
    }

    override fun deleteData(id: Int) {
        val db = this.writableDatabase;
        db.delete(TABLE_HABITS_DETAILS, "$KEY_ID = ?", arrayOf(id.toString()))
        db.close()
    }

    override fun deleteAllData() {
        val db = this.writableDatabase
        db.delete(TABLE_HABITS_DETAILS, null, null)
        db.close()
    }
}